mtags/src/main/scala-2/scala/tools/nsc/interactive/GlobalProxy.scala

```diff
@@ -49,6 +50,7 @@ trait GlobalProxy { this: MetalsGlobal =>
           )
       }
     }
+    this.scheduler = new WorkScheduler
```

Comment:
So I guess I don't fully understand this part. What exactly does create and assigning a new WorkScheduler do? Mainly how does it help this situation?

Reply:
This is a bit of a hack. When we shutdown the compiler the scheduler gets replaced with something like `NoWorkScheduler` which will not accept any work. We replace it back with the default one so that the modified thread can work with it.

Reply:
🤔 ok, just to make sure I understand, after we shutdown, and then right below when we create a new `MetalsGlobalThread`, don't we also get a new compiler, which would already have a new WorkSchedule?

Reply:
Whenever Global is created it comes with it's own thread and WorkScheduler, so we can't create a new one. Here, we shutdown the compiler, which shuts down the thread and replaces the scheduler. After that, we replace the scheduler with a working one and add the new thread, in this way hijack this particular Global object.

Reply:
Alright, I think I understand. Thanks a lot for explaining.

Reply:
I did dig into that for a while before @mlachkar suggestion, so I do understand your confusion here. It's not perfect, it would probably be much better to just use the defaults, but I don't want to change too much and the reason for the changed compiler thread might still exist.

Reply:
Yea understandable, just trying to wrap my head around what's fully going on here.